scripts/truffle/bot.js

```diff
@@ -0,0 +1,304 @@
+const { BN } = require("@openzeppelin/test-helpers");
+const { sleep } = require("../../test-utils/utils");
+const { RAY, RAD, ONE_ETH } = require("../../test-utils/constants");
+
+const mcdJSON = require("../../config/mcd_testchain.json");
+const bpJSON = require("../../config/bprotocol_testchain.json");
+const { default: Web3 } = require("web3");
+
+// MCD Contracts
+const DssCdpManager = artifacts.require("DssCdpManager");
+const Dai = artifacts.require("Dai");
+const DaiJoin = artifacts.require("DaiJoin");
+const GemJoin = artifacts.require("GemJoin");
+const WETH9 = artifacts.require("WETH9");
+const Vat = artifacts.require("Vat");
+const Spotter = artifacts.require("Spotter");
+const LiquidatorInfo = artifacts.require("LiquidatorInfo");
+
+// B.Protocol Contracts
+const BCdpManager = artifacts.require("BCdpManager");
+const Pool = artifacts.require("Pool");
+const OSM = artifacts.require("OSM");
+
+let bCdpManager;
+let pool;
+let dssCdpManager;
+let gemJoin;
+let weth;
+let vat;
+
+let spot;
+let liqInfo;
+
+let MEMBER_1 = bpJSON.MEMBER_1;
+
+const ILK_ETH = web3.utils.padRight(web3.utils.asciiToHex("ETH-A"), 64);
+
+const topped = new Map();
+const bitten = new Map();
+const pending = new Map();
+
+module.exports = async function (callback) {
+  bCdpManager = await BCdpManager.at(bpJSON.B_CDP_MANAGER);
+  pool = await Pool.at(bpJSON.POOL);
+  dai = await Dai.at(mcdJSON.MCD_DAI);
+  daiJoin = await DaiJoin.at(mcdJSON.MCD_JOIN_DAI);
+  dssCdpManager = await DssCdpManager.at(mcdJSON.CDP_MANAGER);
+  gemJoin = await GemJoin.at(mcdJSON.MCD_JOIN_ETH_A);
+  vat = await Vat.at(mcdJSON.MCD_VAT);
+  osm = await OSM.at(mcdJSON.PIP_ETH);
+  spot = await Spotter.at(mcdJSON.MCD_SPOT);
+  const gem = await gemJoin.gem();
+  weth = await WETH9.at(gem);
+  liqInfo = await LiquidatorInfo.at(bpJSON.FLATLIQUIDATOR_INFO);
+
+  // initialize
+  await init();
+
+  // const subscription = web3.eth.subscribe(
+  //   "logs",
+  //   {
+  //     address: spot.address,
+  //   },
+  //   async (error, result) => {
+  //     try {
+  //       await processUntop();
+  //       await processBite();
+  //     } catch (err) {
+  //       console.log(err);
+  //     }
+  //   }
+  // );
+
+  web3.eth.subscribe("newBlockHeaders", async (error, event) => {
+    try {
+      if (!error) {
+        console.log("Block: " + event.number);
+        await processCdps();
+      } else {
+        console.log(error);
+      }
+    } catch (err) {
+      console.log(err);
+    }
+  });
+
+  while (true) {
+    await sleep(10);
+  }
+};
+
+async function processUntop(cdp) {
+  try {
+    // isToppedUp && !isBitten && isSafe
+    const toppedUp = await topped.get(cdp);
```

Comment:
no need to make this call. should untop will return true, only if already topped.